### Task 1

# Data Quality Assessment
######  Assessmentn of the data quality and completeness in preparation for analysis 

> Datasets contains four tables 
- Transactions 
- CustomerAddress 
- NewCustorList
- CustomerDemographic

###  Standard Data Qulaity Dimensions 

- Correct Values | Accuracy 
- Data Fields with Values | Completeness 
- Values Free from Contradiction | Consistency 
- Data Items with Value Meta-data | Relevancy 
- Data Containing Allowable Values | Validity 
- Records that are Duplicated | Uniqueness 



### Import libraries 

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns 

### Read the Dataset

In [2]:
data_file = pd.ExcelFile("KPMG_VI_New_raw_data_update_final.xlsx")

In [3]:
trans = pd.read_excel(data_file, sheet_name=1, header=1)
c_list = pd.read_excel(data_file, sheet_name=2, header=1)
c_Demo = pd.read_excel(data_file, sheet_name=3, header=1)
c_Add = pd.read_excel(data_file, sheet_name=4, header=1)

### Transactions 

- 20,000 rows and 13 columns 
- Total 1,542 missing values in dataset
- 3494 unique customer_id 
- transaction in dataset from 2017-01-01 to 2017-12-31

In [4]:
trans.head()

transaction_id  product_id  ...  standard_cost product_first_sold_date
0               1           2  ...          53.62                 41245.0
1               2           3  ...         388.92                 41701.0
2               3          37  ...         248.82                 36361.0
3               4          88  ...         381.10                 36145.0
4               5          78  ...         709.48                 42226.0

[5 rows x 13 columns]

In [5]:
trans.shape

(20000, 13)

In [6]:
trans.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   transaction_id           20000 non-null  int64         
 1   product_id               20000 non-null  int64         
 2   customer_id              20000 non-null  int64         
 3   transaction_date         20000 non-null  datetime64[ns]
 4   online_order             19640 non-null  float64       
 5   order_status             20000 non-null  object        
 6   brand                    19803 non-null  object        
 7   product_line             19803 non-null  object        
 8   product_class            19803 non-null  object        
 9   product_size             19803 non-null  object        
 10  list_price               20000 non-null  float64       
 11  standard_cost            19803 non-null  float64       
 12  product_first_sold_date  19803 n

In [7]:
trans.isnull().sum()

transaction_id               0
product_id                   0
customer_id                  0
transaction_date             0
online_order               360
order_status                 0
brand                      197
product_line               197
product_class              197
product_size               197
list_price                   0
standard_cost              197
product_first_sold_date    197
dtype: int64

In [8]:
trans.customer_id.value_counts()

2183    14
1068    14
2476    14
2072    13
1672    13
        ..
2395     1
1569     1
1203     1
2271     1
1544     1
Name: customer_id, Length: 3494, dtype: int64

In [9]:
trans.describe()

transaction_id   product_id  ...  standard_cost  product_first_sold_date
count    20000.000000  20000.00000  ...   19803.000000             19803.000000
mean     10000.500000     45.36465  ...     556.046951             38199.776549
std       5773.647028     30.75359  ...     405.955660              2875.201110
min          1.000000      0.00000  ...       7.210000             33259.000000
25%       5000.750000     18.00000  ...     215.140000             35667.000000
50%      10000.500000     44.00000  ...     507.580000             38216.000000
75%      15000.250000     72.00000  ...     795.100000             40672.000000
max      20000.000000    100.00000  ...    1759.850000             42710.000000

[8 rows x 7 columns]

In [10]:
trans.transaction_date.sort_values()

516     2017-01-01
5876    2017-01-01
3459    2017-01-01
12484   2017-01-01
19130   2017-01-01
           ...    
605     2017-12-30
15269   2017-12-30
19906   2017-12-30
15756   2017-12-30
12003   2017-12-30
Name: transaction_date, Length: 20000, dtype: datetime64[ns]

In [11]:
trans.duplicated().any()

False

In [12]:
trans.columns

Index(['transaction_id', 'product_id', 'customer_id', 'transaction_date',
       'online_order', 'order_status', 'brand', 'product_line',
       'product_class', 'product_size', 'list_price', 'standard_cost',
       'product_first_sold_date'],
      dtype='object')

In [13]:
trans.product_class.value_counts()

medium    13826
high       3013
low        2964
Name: product_class, dtype: int64

In [14]:
trans.order_status.value_counts()

Approved     19821
Cancelled      179
Name: order_status, dtype: int64

In [15]:
trans.brand.value_counts()

Solex             4253
Giant Bicycles    3312
WeareA2B          3295
OHM Cycles        3043
Trek Bicycles     2990
Norco Bicycles    2910
Name: brand, dtype: int64

In [16]:
trans.product_line.value_counts()

Standard    14176
Road         3970
Touring      1234
Mountain      423
Name: product_line, dtype: int64

### CustomerList

- 1000 rows, 23 columns 
- 317 missing values in dataset
- many customer has incorrect DOB 
- Duplicate columns `Unnamed:20` and  `Rank`


In [17]:
c_list.head()

first_name  last_name  gender  ...  Unnamed: 20 Rank     Value
0    Chickie    Brister    Male  ...            1    1  1.718750
1      Morly     Genery    Male  ...            1    1  1.718750
2    Ardelis  Forrester  Female  ...            1    1  1.718750
3     Lucine      Stutt  Female  ...            4    4  1.703125
4    Melinda     Hadlee  Female  ...            4    4  1.703125

[5 rows x 23 columns]

In [44]:
c_list.shape

(1000, 23)

In [18]:
c_list.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 23 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   first_name                           1000 non-null   object        
 1   last_name                            971 non-null    object        
 2   gender                               1000 non-null   object        
 3   past_3_years_bike_related_purchases  1000 non-null   int64         
 4   DOB                                  983 non-null    datetime64[ns]
 5   job_title                            894 non-null    object        
 6   job_industry_category                835 non-null    object        
 7   wealth_segment                       1000 non-null   object        
 8   deceased_indicator                   1000 non-null   object        
 9   owns_car                             1000 non-null   object        
 10  tenure       

In [20]:
c_list.isnull().sum().sum()

317

In [21]:
c_list.describe()

past_3_years_bike_related_purchases  ...        Value
count                          1000.000000  ...  1000.000000
mean                             49.836000  ...     0.881714
std                              27.796686  ...     0.293525
min                               0.000000  ...     0.340000
25%                              26.750000  ...     0.649531
50%                              51.000000  ...     0.860000
75%                              72.000000  ...     1.075000
max                              99.000000  ...     1.718750

[8 rows x 11 columns]

In [22]:
c_list.job_title.value_counts()

Associate Professor             15
Environmental Tech              14
Software Consultant             14
Chief Design Engineer           13
VP Sales                        12
                                ..
Budget/Accounting Analyst II     1
Accounting Assistant II          1
Programmer IV                    1
Account Representative II        1
Safety Technician I              1
Name: job_title, Length: 184, dtype: int64

In [27]:
c_list.job_industry_category.value_counts()

Financial Services    203
Manufacturing         199
Health                152
Retail                 78
Property               64
IT                     51
Entertainment          37
Argiculture            26
Telecommunications     25
Name: job_industry_category, dtype: int64

In [23]:
c_list.duplicated().any()

False

In [25]:
c_list.gender.value_counts()

Female    513
Male      470
U          17
Name: gender, dtype: int64

In [26]:
c_list.country.value_counts()

Australia    1000
Name: country, dtype: int64

In [48]:
c_list.DOB.sort_values().sample(10)

338   1956-06-19
933   1975-01-19
856   1953-03-27
971   1999-03-04
452   1943-04-06
306   1954-11-09
665   1960-01-21
88    1950-02-10
804   1947-07-13
137   1967-09-27
Name: DOB, dtype: datetime64[ns]

In [43]:
postcode = np.sort(c_list.postcode.unique())

In [44]:
postcode

array([2000, 2007, 2009, 2010, 2011, 2015, 2016, 2017, 2019, 2022, 2024,
       2025, 2026, 2028, 2030, 2031, 2032, 2033, 2034, 2035, 2036, 2038,
       2042, 2046, 2047, 2049, 2050, 2062, 2066, 2067, 2068, 2069, 2070,
       2071, 2072, 2073, 2074, 2075, 2076, 2077, 2082, 2088, 2089, 2090,
       2092, 2093, 2096, 2097, 2099, 2101, 2103, 2104, 2107, 2110, 2111,
       2112, 2113, 2114, 2117, 2118, 2120, 2121, 2122, 2125, 2126, 2127,
       2128, 2130, 2134, 2135, 2137, 2138, 2140, 2141, 2142, 2144, 2145,
       2146, 2147, 2148, 2151, 2153, 2154, 2155, 2156, 2158, 2160, 2161,
       2163, 2164, 2165, 2166, 2168, 2170, 2171, 2173, 2176, 2177, 2190,
       2192, 2193, 2194, 2195, 2196, 2197, 2199, 2200, 2203, 2205, 2207,
       2209, 2210, 2211, 2212, 2213, 2217, 2218, 2219, 2220, 2221, 2222,
       2223, 2224, 2226, 2227, 2228, 2230, 2231, 2232, 2233, 2234, 2250,
       2251, 2256, 2257, 2258, 2259, 2261, 2262, 2263, 2280, 2281, 2284,
       2287, 2289, 2290, 2291, 2292, 2298, 2300, 23

In [45]:
c_list.owns_car.value_counts()

No     507
Yes    493
Name: owns_car, dtype: int64

### CustomerDemographic

- 4000 rows, 13 columns 
- 1763 missing values in dataset
- uncleared data into column `default`
- incorrect DOB's in dataset

In [18]:
c_Demo.head()

customer_id      first_name  ... owns_car tenure
0            1         Laraine  ...      Yes   11.0
1            2             Eli  ...      Yes   16.0
2            3           Arlin  ...      Yes   15.0
3            4          Talbot  ...       No    7.0
4            5  Sheila-kathryn  ...      Yes    8.0

[5 rows x 13 columns]

In [50]:
c_Demo.shape

(4000, 13)

In [51]:
c_Demo.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 13 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   customer_id                          4000 non-null   int64         
 1   first_name                           4000 non-null   object        
 2   last_name                            3875 non-null   object        
 3   gender                               4000 non-null   object        
 4   past_3_years_bike_related_purchases  4000 non-null   int64         
 5   DOB                                  3913 non-null   datetime64[ns]
 6   job_title                            3494 non-null   object        
 7   job_industry_category                3344 non-null   object        
 8   wealth_segment                       4000 non-null   object        
 9   deceased_indicator                   4000 non-null   object        
 10  default     

In [53]:
c_Demo.isnull().sum().sum()

1763

In [56]:
c_Demo.duplicated("customer_id").any()

False

In [57]:
c_Demo.owns_car.value_counts()

Yes    2024
No     1976
Name: owns_car, dtype: int64

In [60]:
c_Demo.DOB.sort_values()

33     1843-12-21
719    1931-10-23
1091   1935-08-22
3409   1940-09-22
2412   1943-08-11
          ...    
3778          NaT
3882          NaT
3930          NaT
3934          NaT
3997          NaT
Name: DOB, Length: 4000, dtype: datetime64[ns]

In [61]:
c_Demo.wealth_segment.value_counts()

Mass Customer        2000
High Net Worth       1021
Affluent Customer     979
Name: wealth_segment, dtype: int64

### CustomerAddress 

- 3999 rows, 6 columns 
- no duplicated data 
- Inconsistent values for the same attribute in State columns

In [62]:
c_Add.head()

customer_id              address  ...    country property_valuation
0            1   060 Morning Avenue  ...  Australia                 10
1            2  6 Meadow Vale Court  ...  Australia                 10
2            4   0 Holy Cross Court  ...  Australia                  9
3            5  17979 Del Mar Point  ...  Australia                  4
4            6     9 Oakridge Court  ...  Australia                  9

[5 rows x 6 columns]

In [64]:
c_Add.shape

(3999, 6)

In [63]:
c_Add.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3999 entries, 0 to 3998
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   customer_id         3999 non-null   int64 
 1   address             3999 non-null   object
 2   postcode            3999 non-null   int64 
 3   state               3999 non-null   object
 4   country             3999 non-null   object
 5   property_valuation  3999 non-null   int64 
dtypes: int64(3), object(3)
memory usage: 187.6+ KB


In [75]:
c_Add.describe()

customer_id     postcode  property_valuation
count  3999.000000  3999.000000         3999.000000
mean   2003.987997  2985.755939            7.514379
std    1154.576912   844.878364            2.824663
min       1.000000  2000.000000            1.000000
25%    1004.500000  2200.000000            6.000000
50%    2004.000000  2768.000000            8.000000
75%    3003.500000  3750.000000           10.000000
max    4003.000000  4883.000000           12.000000

In [77]:
c_Add.duplicated("customer_id").any()

False

In [78]:
c_Add.country.value_counts()

Australia    3999
Name: country, dtype: int64

In [79]:
c_Add.state.value_counts()

NSW                2054
VIC                 939
QLD                 838
New South Wales      86
Victoria             82
Name: state, dtype: int64